topology/probes/vpp/vpp.go

```diff
@@ -103,268 +124,508 @@ func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
+	return node
 }
 
-func (p *Probe) createOrUpdateInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails) *graph.Node {
-	metadata := graph.Metadata{"IfIndex": int64(intf.SwIfIndex), "Type": "vpp"}
+func (p *Probe) createOrUpdateInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails, metadata graph.Metadata) *graph.Node {
+	vrfID := p.getInterfaceVrfID(ch, intf.SwIfIndex)
+	name := interfaceName(intf.InterfaceName)
 
 	p.Ctx.Graph.Lock()
 	defer p.Ctx.Graph.Unlock()
 
+	metadata["Driver"] = "vpp"
+	metadata["Type"] = "interface"
+	metadata["Name"] = name
+	metadata["IfIndex"] = int64(intf.SwIfIndex)
+	metadata["MAC"] = interfaceMAC(intf.L2Address[:intf.L2AddressLength])
+	metadata["MTU"] = int64(intf.LinkMtu)
+	metadata["Speed"] = int64(intf.LinkSpeed)
+	metadata["State"] = interfaceUpDown(intf.AdminUpDown)
+	metadata["Duplex"] = interfaceDuplex(intf.LinkDuplex)
+	metadata["VrfID"] = vrfID
+
+	if strings.HasPrefix(name, "host-") {
+		metadata["InterfaceName"] = name[5:]
+	}
+
 	var err error
-	node := p.Ctx.Graph.LookupFirstNode(metadata)
+	node := p.getInterface(intf.SwIfIndex)
 	if node == nil {
 		node, err = p.Ctx.Graph.NewNode(graph.GenID(), metadata)
 		if err != nil {
 			p.Ctx.Logger.Error(err)
 			return nil
 		}
+		p.NotifyEvent(graph.NodeAdded, node)
+
 		if _, err = p.Ctx.Graph.Link(p.vppRootNode, node, topology.OwnershipMetadata()); err != nil {
 			p.Ctx.Logger.Error(err)
 			return nil
 		}
+	} else {
+		p.NotifyEvent(graph.NodeUpdated, node)
 	}
-
-	tr := p.Ctx.Graph.StartMetadataTransaction(node)
-	defer tr.Commit()
-	tr.AddMetadata("Driver", "vpp")
-	tr.AddMetadata("Name", strings.Trim(string(intf.InterfaceName), "\000"))
-	tr.AddMetadata("IfIndex", int64(intf.SwIfIndex))
-	tr.AddMetadata("MAC", interfaceMAC(intf.L2Address[:intf.L2AddressLength]))
-	tr.AddMetadata("MTU", int64(intf.LinkMtu))
-	tr.AddMetadata("Speed", int64(intf.LinkSpeed))
-	state := interfaceUpDown(intf.AdminUpDown)
-	tr.AddMetadata("State", state)
-	if state != "DOWN" {
-		tr.AddMetadata("Duplex", interfaceDuplex(intf.LinkDuplex))
-	}
-	tr.AddMetadata("VrfID", p.getInterfaceVrfID(ch, intf.SwIfIndex))
-
 	return node
 }
 
 func interfaceNeedUpdate(i1, i2 *interfaces.SwInterfaceDetails) bool {
-	if (i1.LinkMtu != i2.LinkMtu) ||
-		(i1.LinkSpeed != i2.LinkSpeed) ||
-		(i1.LinkDuplex != i2.LinkDuplex) {
-		return true
-	}
-	return false
+	return i1.LinkMtu != i2.LinkMtu ||
+		i1.LinkSpeed != i2.LinkSpeed ||
+		i1.LinkDuplex != i2.LinkDuplex
 }
 
-func (p *Probe) eventAddInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails) {
-	p.createOrUpdateInterface(ch, intf)
+func (p *Probe) eventAddInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails, metadata graph.Metadata) *graph.Node {
+	return p.createOrUpdateInterface(ch, intf, metadata)
 }
 
 func (p *Probe) eventDelInterface(node *graph.Node) {
-	p.Ctx.Graph.Lock()
-	defer p.Ctx.Graph.Unlock()
 	if err := p.Ctx.Graph.DelNode(node); err != nil {
+		p.NotifyEvent(graph.NodeDeleted, node)
```

Comment:
same here, why sending a event ? why not subscribing to the graph ?

Reply:
Small optimization explained in response above